### PR TITLE
Initial LvmCluster controller 

### PR DESCRIPTION
**Pull request by nbalacha**
Initial LvmCluster controller implementation and tests.

### CODE REVIEW

File: api/v1alpha1/lvmcluster_types.go [hunk 1]
```diff
@@ -65,10 +65,15 @@ type DeviceSelector struct {
...
 // }
+
 // LVMClusterStatus defines the observed state of LVMCluster
 type LVMClusterStatus struct {
 	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
 	// Important: Run "make" to regenerate code after modifying this file
+
+	// ready describes if the LvmCluster is ready.
+	// +optional
+	Ready bool `json:"ready,omitempty"`
```
Comment (nbalacha) [inline api/v1alpha1/lvmcluster_types.go:76]: This is just a placeholder field in the status. We will probably end up removing this.

File: controllers/defaults.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+)
+
+var (
+	DefaultValMap = map[string]string{
+		"OPERATOR_NAMESPACE":    "openshift-storage",
+		"TOPOLVM_CSI_IMAGE":     "quay.io/topolvm/topolvm:0.10.3",
+		"CSI_REGISTRAR_IMAGE":   "k8s.gcr.io/sig-storage/csi-node-driver-registrar:v2.3.0",
+		"CSI_PROVISIONER_IMAGE": "k8s.gcr.io/sig-storage/csi-provisioner:v3.0.0",
+		"CSI_ATTACHER_IMAGE":    "k8s.gcr.io/sig-storage/csi-attacher:v3.3.0",
+		"CSI_SNAPSHOTTER_IMAGE": "k8s.gcr.io/sig-storage/csi-snapshotter:v4.2.0",
```
Comment (leelavg) [inline controllers/defaults.go:30]: - we'll never need attacher and snapshotter isn't being used as of now

**Review by leelavg — COMMENTED** [review]
minor comment, lgtm.

File: controllers/defaults.go (hunk 2 above)
Comment (nbalacha) [reply controllers/defaults.go:30]: The topolvm makefile is using the csi attacher which is why I added it:
```EXTERNAL_PROVISIONER_SRC  = $(SRC_ROOT)/external-provisioner
NODE_DRIVER_REGISTRAR_SRC = $(SRC_ROOT)/node-driver-registrar
EXTERNAL_ATTACHER_SRC     = $(SRC_ROOT)/external-attacher
EXTERNAL_RESIZER_SRC      = $(SRC_ROOT)/external-resizer
LIVENESSPROBE_SRC         = $(SRC_ROOT)/livenessprobe```

Comment (leelavg) [reply controllers/defaults.go:30]: - there's no separate liveness probe iirc, same hypertopolvm handles that

Comment (leelavg) [reply controllers/defaults.go:30]: @nbalacha 

1. pls add the liveness probe
2. attacher can be [removed](https://github.com/topolvm/topolvm/blob/6697490595efd7cfa78bffc6b7a0581b33c18bd2/CHANGELOG.md#changed-3)

Comment (nbalacha) [reply controllers/defaults.go]: Done.

**Review by leelavg — APPROVED** [review]
/lgtm

File: controllers/defaults.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+*/
+
+package controllers
+
+import (
+	"os"
+)
+
+var (
+	DefaultValMap = map[string]string{
```
Comment (sp98) [inline controllers/defaults.go:24]:
```suggestion
	defaultValMap = map[string]string{
```
Looks like it won't be used outside this package.

File: controllers/lvmcluster_controller.go [hunk 4]
```diff
@@ -56,81 +60,135 @@ type LVMClusterReconciler struct {
...
+		if contains(instance.GetFinalizers(), lvmClusterFinalizer) {
+			for _, unit := range resourceList {
+				err := unit.ensureDeleted(r, ctx, instance)
+				if err != nil {
+					return ctrl.Result{}, fmt.Errorf("failed cleaning up: %s %w", unit.getName(), err)
+				}
+			}
+			instance.ObjectMeta.Finalizers = remove(instance.ObjectMeta.Finalizers, lvmClusterFinalizer)
+			if err := r.Client.Update(context.TODO(), instance); err != nil {
+				r.Log.Info("Failed to remove finalizer from LvmCLuster", "LvmCluster", instance.Name)
```
Comment (sp98) [inline controllers/lvmcluster_controller.go:111]:
```suggestion
				r.Log.Info("failed to remove finalizer from LvmCLuster", "LvmCluster", instance.Name)
```

File: controllers/lvmcluster_controller.go [hunk 5]
```diff
@@ -56,81 +60,135 @@ type LVMClusterReconciler struct {
...
-	// handle create/update
-	for _, unit := range unitList {
-		err := unit.ensureCreated(r, ctx, *lvmCluster)
-		if err != nil {
-			return result, fmt.Errorf("failed reconciling: %s %w", unit.getName(), err)
+	if !contains(instance.GetFinalizers(), lvmClusterFinalizer) {
+		r.Log.Info("Finalizer not found for LvmCluster. Adding finalizer.", "LvmCluster", instance.Name)
+		instance.ObjectMeta.Finalizers = append(instance.ObjectMeta.Finalizers, lvmClusterFinalizer)
+		if err := r.Client.Update(context.TODO(), instance); err != nil {
+			r.Log.Info("Failed to update LvmCluster with finalizer.", "LvmCluster", instance.Name)
```
Comment (sp98) [inline controllers/lvmcluster_controller.go:122]:
```suggestion
			r.Log.Info("failed to update LvmCluster with finalizer.", "LvmCluster", instance.Name)
```

**Review by sp98 — CHANGES_REQUESTED** [review]
nit

File: controllers/defaults.go (hunk 3 above)
Comment (nbalacha) [reply controllers/defaults.go]: Done

File: controllers/lvmcluster_controller.go (hunk 4 above)
Comment (nbalacha) [reply controllers/lvmcluster_controller.go]: Done

File: controllers/lvmcluster_controller.go (hunk 5 above)
Comment (nbalacha) [reply controllers/lvmcluster_controller.go]: Done

**Review by sp98 — APPROVED** [review]

**Pull request merged**